application/common/models/User.php

```diff
@@ -292,6 +295,110 @@ private function validatePassword(): Closure
         };
     }
 
+    /*
+     * Check the following to decide if HIBP should be called:
+     *  - if Yii::$app->params['hibpCheckOnLogin'] is true
+     *  - if $this->password is not empty
+     *
+     */
+    public function shouldHibpBeChecked(): bool
+    {
+        return (\Yii::$app->params['hibpCheckOnLogin'] &&
+                ! empty($this->password) &&
+                time() >= strtotime($this->currentPassword->check_hibp_after) &&
+                $this->currentPassword->hibp_is_pwned == 'no');
+    }
+
+    /*
+     * If user is due to have password checked with HIBP, check it
+     * If password is found to be pwned, process it
+     * Fail gracefully to allow user to login if HIBP is unavailable
+     */
+    public function checkAndProcessHIBP(): void
+    {
+        if (! $this->shouldHibpBeChecked()) {
+            return;
+        }
+
+        if (! $this->isPasswordPwned()) {
+            $this->currentPassword->extendHibpCheckAfter();
+            return;
+        }
+
+        $this->trackPwnedPasswordGAEvent();
+
+        if (\Yii::$app->params['hibpTrackingOnly']) {
+            // extend check after date to only track user once per checking period
+            $this->currentPassword->extendHibpCheckAfter();
+            return;
+        }
+
+        $this->currentPassword->markPwned();
+
+        // notify user
+        try {
+            /* @var $emailer Emailer */
+            $emailer = \Yii::$app->emailer;
+            $emailer->sendMessageTo(EmailLog::MESSAGE_TYPE_PASSWORD_PWNED, $this, [
+                'bccAddress' => \Yii::$app->params['hibpNotificationBcc']
```

Comment:
When in "track-only" mode, were we supposed to also send the email to the `hibpNotificationBcc` recipient?

For some reason I thought so, but perhaps "track-only" was merely intended to gather GA analytics, not emails to the IT support staff about which accounts are using a pwned password. Or does the GA analytics call individually identify the account whose password was leaked?

Reply:
tracking only just gets numbers. it is intended just to get a sense of scale of the problem before support is possibly overwhelmed.